src/pgrn-custom-scan.c

```diff
@@ -19,3 +77,102 @@ PGrnCustomScanDisable(void)
 {
 	PGrnCustomScanEnabled = false;
 }
+
+static void
+PGrnSetRelPathlistHook(PlannerInfo *root,
+					   RelOptInfo *rel,
+					   Index rti,
+					   RangeTblEntry *rte)
+{
+	CustomPath *cpath;
+	if (PreviousSetRelPathlistHook)
+	{
+		PreviousSetRelPathlistHook(root, rel, rti, rte);
+	}
+
+	if (!PGrnCustomScanGetEnabled())
+	{
+		return;
+	}
+
+	cpath = makeNode(CustomPath);
+	cpath->path.pathtype = T_CustomScan;
+	cpath->path.parent = rel;
+	cpath->path.pathtarget = rel->reltarget;
+	cpath->methods = &PGrnPathMethods;
+
+	add_path(rel, &cpath->path);
+}
+
+static Plan *
+PGrnPlanCustomPath(PlannerInfo *root,
+				   RelOptInfo *rel,
+				   struct CustomPath *best_path,
+				   List *tlist,
+				   List *clauses,
+				   List *custom_plans)
+{
+	CustomScan *cscan = makeNode(CustomScan);
+	cscan->methods = &PGrnScanMethods;
+	return &(cscan->scan.plan);
+}
+
+static List *
+PGrnReparameterizeCustomPathByChild(PlannerInfo *root,
+									List *custom_private,
+									RelOptInfo *child_rel)
+{
+	return NIL;
+}
+
+static Node *
+PGrnCreateCustomScanState(CustomScan *cscan)
+{
+	PGrnScanState *state =
+		(PGrnScanState *) newNode(sizeof(PGrnScanState), T_CustomScanState);
+
+	state->parent.methods = &PGrnExecuteMethods;
+
+	return (Node *) &(state->parent);
+}
+
+static void
+PGrnBeginCustomScan(CustomScanState *cscanstate, EState *estate, int eflags)
+{
+}
+
+static TupleTableSlot *
+PGrnExecCustomScan(CustomScanState *node)
+{
+	return NULL;
+}
+
+static void
+PGrnExplainCustomScan(CustomScanState *node, List *ancestors, ExplainState *es)
+{
+	ExplainPropertyText("PGroongaScan", "DEBUG", es);
+}
+
+static void
+PGrnEndCustomScan(CustomScanState *node)
+{
+}
+
+static void
+PGrnReScanCustomScan(CustomScanState *node)
+{
+}
+
+void
+PGrnInitializeCustomScan(void)
+{
+	PreviousSetRelPathlistHook = set_rel_pathlist_hook;
+	set_rel_pathlist_hook = PGrnSetRelPathlistHook;
+
+	RegisterCustomScanMethods(&PGrnScanMethods);
+}
+
+void
+PGrnFinalizeCustomScan(void)
+{
+}
```

Comment:
Do we need `set_rel_pathlist_hook = PreviousSetRelPathlistHook` here?

Reply:
I could not find any documentation if `set_rel_pathlist_hook = PreviousSetRelPathlistHook` is needed or not.

However, there is no UnregisterCustomScanMethods() pair to RegisterCustomScanMethods(), and registered CustomScanMethods cannot be reset.
https://github.com/postgres/postgres/blob/cb36f8ec2124b0c495da057a2f55735fe858be0e/src/include/nodes/extensible.h#L160
So I have decided that I don't need to reset the `set_rel_pathlist_hook` too.

Reply:
Hmm. We had some crash bugs that are caused by running some registered functions are used after PGroonga is finalized.

Can we ensure disabling all registered functions (exits immediately?) after PGroonga is finalized?

Reply:
I got it.
I will implement function disabling in a new PR.